Add playbooks and workflow job to release 6.3, 6.2 and Nightly

ehelms opened the pull request.

ehelms commented:
This is a bit of a work in progress as its adding the following that is worth consideration (and I plan to add a few more things to the workflow):

 * introduces Ansible base configuration
 * adds playbooks directory for storing playbooks to be used during a job similar to scripts
 * adds the use of the Ansible jenkins plugin (https://github.com/SatelliteQE/robottelo-ci/tree/master/jobs/release)

rochacbruno commented on playbooks/6.3/sync_repositories.yml:
```diff
@@ -0,0 +1,69 @@
+        module: katello
+        username: "{{ username }}"
+        password: "{{ password }}"
+        server_url: "{{ server }}"
```
I am wondering if katello module cannot read username, password and server_url from the upper level global `vars` instead of redefining the same vars for each task?

ehelms replied:
Not sure I follow entirely -- here this isn't redefining the variables, but rather declaring them for use in the module. I don't know that its possible to define 'global' module parameters given how modules work in Ansible. One thing that can be done is to re-factor things like this:

```
katello.yml
---
- name: "{{ name }}"
  local_action:
    module: katello
    username: "{{ username }}"
    password: "{{ password }}"
    server_url: "{{ server }}"
    entity: "{{ entity }}"
    action: "{{ action }}"
    params: "{{ params }}"
~                          

sync_repositories.yml
---
- include: katello.yml
  vars:
    entity: 'repository'
    action: 'sync'
    params:
      name: 'Blah'
      organization: "{{ organization }}"
```

rochacbruno replied:
For some core modules it works

```yml
group_vars/all
----
katello_username = admin
katello_password = s3c43t
```

```yml
roles/rolenaname/tasks/main.yml
---
- name: something
  local_action:
      module: katello
      entity: SomeEntity
      action: create
      params: {}
      # username and password if not passed here will be read from group_vars/all (following ansible variable precedence) but it depends on how module has been written.
```

ehelms replied:
Does this only work for roles or can these be defined globally for playbooks?

rochacbruno replied:
It can be defined globally in inventory file http://docs.ansible.com/ansible/playbooks_variables.html#variables-defined-in-inventory

where `[all]` section stands for every host so `vars:` there is available for every context.

ehelms commented:
Added timing plugin to the ansible.cfg to get timed outputs, e.g.:

```
PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
Thursday 29 September 2016  16:01:36 -0400 (0:00:00.065)       0:00:00.065 **** 
ok: [localhost]

TASK [Sync Foreman 1.13] *******************************************************
Thursday 29 September 2016  16:01:47 -0400 (0:00:10.315)       0:00:10.381 **** 
changed: [localhost -> localhost]

TASK [Sync Foreman Plugins 1.13] ***********************************************
Thursday 29 September 2016  16:02:04 -0400 (0:00:17.831)       0:00:28.213 **** 
changed: [localhost -> localhost]

TASK [Sync Katello 3.2] ********************************************************
Thursday 29 September 2016  16:02:18 -0400 (0:00:13.876)       0:00:42.089 **** 
changed: [localhost -> localhost]

TASK [Sync Katello 3.2 - Pulp] *************************************************
Thursday 29 September 2016  16:02:33 -0400 (0:00:14.204)       0:00:56.294 **** 
changed: [localhost -> localhost]

TASK [Sync Katello 3.2 - Candlepin] ********************************************
Thursday 29 September 2016  16:02:46 -0400 (0:00:13.409)       0:01:09.704 **** 
changed: [localhost -> localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=6    changed=5    unreachable=0    failed=0   

Thursday 29 September 2016  16:03:00 -0400 (0:00:13.981)       0:01:23.686 **** 
=============================================================================== 
Sync Foreman 1.13 ------------------------------------------------------ 17.83s
Sync Katello 3.2 ------------------------------------------------------- 14.20s
Sync Katello 3.2 - Candlepin ------------------------------------------- 13.98s
Sync Foreman Plugins 1.13 ---------------------------------------------- 13.88s
Sync Katello 3.2 - Pulp ------------------------------------------------ 13.41s
setup ------------------------------------------------------------------ 10.32s
```

omaciel commented:
ping?

rochacbruno commented:
NOTE: @omaciel @ehelms the robottelo.populate proposed in https://github.com/SatelliteQE/robottelo/pull/4132 has the same idiom as Ansible and with a few modifications can be used as Ansible module.

It uses nailgun currently, and plan is to add CLI compat layer.

the advantages of using that is that some special cases are threat in the module. e.g: Not all nailgun entities can be searched by attributes, some only by search_query (I guess it is due to API inconsistency) but I fixed that cases using a rules dictionary.

Also robottelo.populate can look to existing entities to avoid duplication or error.

ehelms commented:
Couple of questions:

 1) When you say module, you mean in the python module sense not the Ansible module sense of the term?
 2) Why build out something that acts and looks like Ansible instead of just building and using an Ansible module (or contributing the existing ecosystem in this space) ?
 3) Seems we are adding yet another item in this space without the benefits of community (i.e. there is the Ansible module I started, there is also https://github.com/adfinis-sygroup/foreman-yml)

rochacbruno commented:
@ehelms 

> 1. When you say module, you mean in the python module sense not the Ansible module sense of the term?

I say the robottelo.populate module, which is not Ansible module yet, but can be easily integrated, and called by ansible.

> 2. Why build out something that acts and looks like Ansible instead of just building and using an Ansible module (or contributing the existing ecosystem in this space) ?

- It looks exactly like ansible because it should be integrated with ansible, so the idiom is very close.
- It includes some nailgun specific rules and workarounds
- it should integrates with robottelo to be used as a decorator in test_cases, I cannot do that if it was pure ansible module. (but as I said, it can be integrated to be called by ansible)
- it should support **validate** which does not create anything only checks existence

> 3. Seems we are adding yet another item in this space without the benefits of community (i.e. there is the Ansible module I started, there is also https://github.com/adfinis-sygroup/foreman-yml)

- I dont think it is the same purpose, but they can be complements
- robottelo.populate should be available as a decorator in test_cases and has workrounds for nailgun (search) and we need flexibility to configure special cases (e.g: force entity to be populated via CLI)
- The **foreman-yml** is very nice, but the purpose is only for Config not other entities, 
robottelo.populate suports every nailgun entities

plus: I was not aware of other modules + I need to leran more about Ansible modules + It started as a simple "populate and validate" and growed naturally to what I proposed :)

We can move the robottelo.populate to a separate repo and use as standalone tool if needed.

ehelms commented:
Without digging into the code, a few follow up questions:

 1) How tightly coupled is this to robottelo? Does robottelo just need the decorator that is provied by the module or calls into the module?
 2) The description of this makes it sound more like its functionality tied to Nailgun (or perhaps functionality that builds on top of Nailgun)
 3) Based on the above, would you forsee this being its own python module that robottelo requires?

rochacbruno commented:
@ehelms you are right, it is not tight with robottelo, it only must provide decorator to robottelo.

It is based on Nailgun entities, so it is depending on nailgun.

I included a question on PR to get others suggestions, maybe we can turn it in to separate repo `foreman-populate` without the need of robottelo.

zjhuntin commented on workflows/6.3/releaseSatellitePipeline.groovy:
```diff
@@ -0,0 +1,154 @@
+
+                try {
+                    sh "ssh jenkins@${env.LIBVIRT_HOST} \"cd sat-deploy && git -c http.sslVerify=false fetch origin && git reset origin/master --hard\""
+                    sh "ssh jenkins@${env.LIBVIRT_HOST} \"cd sat-deploy/forklift && git -c http.sslVerify=false fetch origin && git reset origin/master --hard\""
```
Same as the last one.

zjhuntin commented on workflows/nightly/releaseSatellitePipeline.groovy:
```diff
@@ -0,0 +1,129 @@
+
+                try {
+                    sh "ssh jenkins@${env.LIBVIRT_HOST} \"cd sat-deploy && git -c http.sslVerify=false fetch ehelms && git reset ehelms/use-forklift-library --hard\""
+                    sh "ssh jenkins@${env.LIBVIRT_HOST} \"cd sat-deploy/forklift && git -c http.sslVerify=false fetch origin && git reset origin/forklift-library --hard\""
```
Here too

zjhuntin commented on workflows/nightly/releaseSatellitePipeline.groovy:
```diff
@@ -0,0 +1,129 @@
+        parallel branches
+}
+
+def runPlaybook(body) {
```
Could we move this function to another file?

zjhuntin replied:
Didn't think about this enough before commenting ignore this! 🎃

ehelms commented:
@evgeni @zjhuntin I think I have addressed all issues and updated this to re-factor out common parts and add some useful functionality. Please review.

evgeni commented on ansible/playbooks/compare_content_views.yml:
```diff
@@ -0,0 +1,47 @@
+---
```
this playbook is not referenced anywhere?

evgeni commented on ansible/playbooks/compare_content_views.yml:
```diff
@@ -0,0 +1,47 @@
+  vars:
+    organization: Sat6-CI
+    server: https://sat-r220-02.lab.eng.rdu2.redhat.com/
+    username: admin
```
user/pass should come from jenkins credentials?

evgeni commented on ansible/playbooks/compare_content_views.yml:
```diff
@@ -0,0 +1,47 @@
+    username: admin
+    password: changeme
+  tasks:
+    - name: 'Compare Satellite 6.3 RHEL7 Test to QA'
```
6.3 is hardcoded?

evgeni commented on ansible/playbooks/create_environment.yml:
```diff
@@ -0,0 +1,16 @@
+---
```
this playbook is not referenced anywhere?

evgeni commented on ansible/playbooks/generate_compose.yml:
```diff
@@ -0,0 +1,12 @@
+        msg: "Generating for RHEL{{ rhel_version }}"
+
+    - name: "Generate compose"
+      shell: 'sh generate-compose rhel{{ rhel_version }} /home/jenkins'
```
s/shell/command/?

ehelms replied:
I think I had to do it this way currently until I extract this functionality out.

evgeni replied:
not sure I can parse your answer :(
my initial comment was that we can use the `command` module instead of the `shell` one, as we do not require a real shell (we don't rely on shell functions)

ehelms replied:
Yea.. and I'm saying that I ended up doing it this way using shell for a reason as far as I can recall. I'd like to just wrap this up into a further additional re-factoring where the compose generation is fully extracted to a playbook and does not rely on this generate-compose script.

evgeni replied:
ACK.

evgeni commented on ansible/playbooks/update_packaging_repo.yml:
```diff
@@ -0,0 +1,11 @@
+    - name: 'Clone packaging repository'
+      git:
+        repo: "https://{{ git_server }}/{{ git_group }}/satellite-packaging.git"
+        dest: "/home/jenkins/satellite-packaging"
```
this means we can't run two sat releases in parallel?

ehelms replied:
Currently, that is correct, the generate compose steps would error if they overlapped. Compose generation could do with some overall re-factoring. I can do that now before this gets merged into SCM or after.

evgeni replied:
the olde jobs have the same limitation, right? then i think its ok to fix that at a later point.

ehelms replied:
They do.

ehelms commented on workflows/releaseSatellitePipeline.groovy:
```diff
@@ -0,0 +1,65 @@
+            }
+        }
+
+    stage "Generate Composes"
```
I guess? Depends how you look at it, currently setup is specific to setting up everything needed to start running the playbooks and executing each subsequent stage. The updating of the packaging repository is entirely specific to generating composes only.

evgeni replied:
Given that we currently don't update the repo at all, the generate compose steps runs with *some* checkout, which means potentially with the wrong or an outdated branch.

So I don't care much when we run the playbook, as long as it runs always before doing composes.

ehelms replied:
Ohh, your comment confused me because that playbook was originally being run. Looks like it got dropped out during the re-factor. Will re-add.

ehelms replied:
Ok should be back now -- https://github.com/SatelliteQE/robottelo-ci/pull/380/files#diff-d82d3a93b4be1bd02f4e2a4632b26087R19

ehelms commented on workflows/lib/runPlaybook.groovy:
```diff
@@ -0,0 +1,61 @@
+
+    for (int i = 0; i < config.items.size(); i++) {
+        def index = i // fresh variable per iteration; i will be mutated
+        branches["${name}-${config.items.get(i)}"] = {
```
I made this configurable now, so that output is little more understandable, e.g.:

```
[install-rhel-7] 
[install-rhel-7] TASK [setup] *******************************************************************
[install-rhel-6]  [WARNING]: provided hosts list is empty, only localhost is available
[install-rhel-6] 
[install-rhel-6] PLAY [localhost] ***************************************************************
[install-rhel-6] 
[install-rhel-6] TASK [setup] *******************************************************************
[install-rhel-7] ok: [localhost]
[install-rhel-7] 
```

Instead of `split0` and `split1` I think this should help readability.

zjhuntin replied:
Agreed, I like that plan!

evgeni commented on workflows/releaseSatellitePipeline.groovy:
```diff
@@ -0,0 +1,91 @@
+
+    stage "Test Installation" {
+
+        runOnLibvirtHost "cd sat-deploy && git -c http.sslVerify=false fetch origin && git reset origin/master --hard"
```
what creates these directories/checkouts first? I don't see any clone in the code.

ehelms replied:
You are correct, at present it is making a couple of assumptions of existence. This whole installation area could use a re-factoring to be done cleaner and probably with 1-2 playbooks. I can do that as part of this change here or later. I personally lean towards later since we use these in production already and it would be nice to get things into master and make incremental updates soonish.

evgeni replied:
ACK

evgeni commented on jobs/release/sat-nightly-release-pipeline.yml:
```diff
@@ -0,0 +1,8 @@
+- job:
```
should this file be called `sat-nightly-release-pipeline.yaml` instead of `sat-nightly-release-pipeline.yml` to be consistent with the other job yamls?

ehelms replied:
Addressed

evgeni commented:
ACK

evgeni commented on workflows/releaseSatellitePipeline.groovy:
```diff
@@ -0,0 +1,91 @@
+        runPlaybook {
+            playbook = 'playbooks/update_packaging_repo.yml'
+            extraVars = [
+                'git_server': env.GIT_HOSTNAME,
```
this needs to be:

```groovy
        def git_hostname = env.GIT_HOSTNAME

        runPlaybook {
            playbook = 'playbooks/update_packaging_repo.yml'
            extraVars = [
				'git_server': git_hostname,
				'git_group': 'satellite6',
				'satellite_version': satellite_version
			]
        }
```

as otherwise it errors out with:
```
java.lang.NullPointerException: Cannot get property 'GIT_HOSTNAME' on null object
```
as @zjhuntin just had

ehelms commented:
@evgeni updated -- good with the current state?
